src/github.com/couchbase/sync_gateway/base/bucket.go

```diff
@@ -121,10 +148,212 @@ func (bucket CouchbaseBucket) StartTapFeed(args sgbucket.TapArguments) (sgbucket
 		}
 		LogTo("Feed", "Using DCP feed for bucket: %q", bucket.GetName())
 		return feed, nil
-	} else {
+
+	case DcpShardFeedType:
+
+		// TODO -- is this guaranteed to _only_ be called once per bucket
+		// throughout lifetime of Sync Gateway?  If not, this will need to get reworked.
+
+		// CBGT initialization
+
+		// Create the TapEvent feed channel that will be passed back to the caller
+		eventFeed := make(chan sgbucket.TapEvent, 10)
+
+		// wrap functions with closures that have references to channel
+		newSyncGatewayPIndexImpl := NewSyncGatewayPIndexImplFactory(eventFeed, bucket, args)
+		openSyncGatewayPIndexImpl := OpenSyncGatewayPIndexImplFactory(eventFeed, bucket, args)
+
+		// register with CBGT
+		cbgt.RegisterPIndexImplType(IndexTypeSyncGateway, &cbgt.PIndexImplType{
+			New:         newSyncGatewayPIndexImpl,
+			Open:        openSyncGatewayPIndexImpl,
+			Count:       nil,
+			Query:       nil,
+			Description: "Sync Gateway Pindex",
+		})
+
+		if err := bucket.initCBGTManager(); err != nil {
+			log.Fatalf("Unable to initialize CBGT for sharded dcp: %v", err)
+		}
+
+		// create the index
+		alreadyExists, err := bucket.checkCBGTIndexExists(bucket.Name)
+		if err != nil {
+			log.Fatalf("Error checking if CBGT index exists: %v", err)
+		}
+		if !alreadyExists {
+			if err := bucket.createCBGTIndex(); err != nil {
+				log.Fatalf("Unable to initialize CBGT index: %v", err)
+			}
+		}
+
+		//  - create a new CBGTDCPFeed and pass in the eventFeed channel
+		feed := &CBGTDCPFeed{
+			eventFeed: eventFeed,
+		}
+		return feed, nil
+
+	default:
 		LogTo("Feed", "Using TAP feed for bucket: %q (based on feed_type specified in config file", bucket.GetName())
 		return bucket.StartCouchbaseTapFeed(args)
+
+	}
+
+}
+
+func (bucket CouchbaseBucket) initCBGTManager() error {
+
+	var username, password string
+
+	feedParams := bucket.spec.FeedParams
+
+	if bucket.spec.Auth != nil {
+		username, password, _ = bucket.spec.Auth.GetCredentials()
+	}
+
+	if username == "" {
+		username = bucket.Name
 	}
+
```

Comment:
Can we reuse our bucket.getDcpAuthHandler() here to do the username="" handling?